### PR TITLE
feat: filter out irrelevant traceback frames [APE-758]

### PR DESCRIPTION
**Pull request by unparalleled-js**
### What I did

fixes: #442 
Fixes: APE-157

* Removes noisy framework traceback frames in `ape console` during `TransactionError` handling
* Removes noisy framework traceback frames in `ape run` during `TransactionError` handling
* Remove noisy framework traceback frames in `ape test` during `TransactionError` handling.

This makes the PR #1337 a lot simpler to follow and smaller.
There is also a lot of benefit to this alone so wanted to include it incase #1337 does not make it in the next release.

### How I did it

 * filter out frames that didn't occur in the local Ape project during the handling of `TransactionError` in CLI, pytest plugin, and IPython plugin.

### How to verify it

Cause unhandled TransactionError (e.g. ContractLogicError) in all 3 environments.

### Checklist

- [ ] All changes are completed
- [ ] New test cases have been added
- [ ] Documentation has been updated

### CODE REVIEW

**Review by fubuloubu — COMMENTED** [review]
This will need some additional QA

File: src/ape/exceptions.py [hunk 1]
```diff
@@ -371,3 +375,38 @@ def __init__(
...
+        return False
+
+    tb = traceback.extract_tb(sys.exc_info()[2])
+    base = str(err.txn.project_manager.path)
+    relevant_tb = [f for f in tb if base in f.filename]
+    if not relevant_tb:
+        return False
+
+    from ape.cli.utils import abort
+    from ape.logging import logger
```
[the remark this answers is not in the document]
Comment (unparalleled-js) [reply src/ape/exceptions.py:404]: yes they are, not sure the best course

**Comment by unparalleled-js** [comment]
was playing around - realizing that it would also be nice to have `SignatureError` do this-- or maybe all `ApeException`

File: src/ape/exceptions.py [hunk 2]
```diff
@@ -371,3 +376,39 @@ def __init__(
...
+
+def handle_ape_exception(err: ApeException, extra_paths: Optional[List[Path]] = None):
+    """
+    Handle a transaction error by showing relevant stack frames,
+    including custom contract frames added to the exception.
+    This method must be called within an ``except`` block or with
+    an exception on the exc-stack.
+
+    Args:
+        err (:class:`~ape.exceptions.TransactionError`): The transaction error
```
Comment (fubuloubu) [inline src/ape/exceptions.py:389]: Change the error class here

File: src/ape/exceptions.py [hunk 3]
```diff
@@ -371,3 +376,39 @@ def __init__(
             kwargs["exception"] = exception
 
         super().__init__(provider, *args, **kwargs)
+
+
+def handle_ape_exception(err: ApeException, extra_paths: Optional[List[Path]] = None):
```
Comment (fubuloubu) [inline src/ape/exceptions.py:381]: Missing `-> bool`

File: src/ape/cli/utils.py [hunk 4]
```diff
@@ -29,3 +31,16 @@ def show(self, file=None):
         """
 
         logger.error(self.format_message())
+
+
+def abort(err: ApeException, show_traceback: Optional[bool] = None) -> Abort:
```
Comment (fubuloubu) [inline src/ape/cli/utils.py:36]: Should this be in `ape.exceptions`?

File: src/ape/exceptions.py [hunk 5]
```diff
@@ -371,3 +376,39 @@ def __init__(
...
+    Returns:
+        bool: ``True`` if outputted something.
+    """
+
+    from ape import project
+    from ape.cli.utils import abort
+    from ape.logging import logger
+
+    tb = traceback.extract_tb(sys.exc_info()[2])
+    allowed_paths = [str(p) for p in [project.path, *(extra_paths or [])]]
```
Comment (fubuloubu) [inline src/ape/exceptions.py:403]: Can `project.path` just be a pass-through argument? e.g. `base_path`

File: src/ape/exceptions.py [hunk 6]
```diff
@@ -371,3 +376,39 @@ def __init__(
...
+        extra_paths (Optional[List[Path]]): Optionally include additional
+          source-path prefixes to use when finding relevant frames.
+
+    Returns:
+        bool: ``True`` if outputted something.
+    """
+
+    from ape import project
+    from ape.cli.utils import abort
+    from ape.logging import logger
```
Comment (fubuloubu) [inline src/ape/exceptions.py:400]: Logging should not create any circular dependencies, right?

File: src/ape/cli/utils.py [hunk 7]
```diff
@@ -1,31 +1,4 @@
...
-
-        super().__init__(message)
-
-    def show(self, file=None):
-        """
-        Override default ``show`` to print CLI errors in red text.
-        """
-
-        logger.error(self.format_message())
+# Kept here for backwards compatibility.
```
Comment (fubuloubu) [inline src/ape/cli/utils.py:3]: Make a note to remove in 0.7

File: src/ape/exceptions.py [hunk 8]
```diff
@@ -371,3 +379,70 @@ def __init__(
...
+
+    def show(self, file=None):
+        """
+        Override default ``show`` to print CLI errors in red text.
+        """
+
+        logger.error(self.format_message())
+
+
+def abort(err: ApeException, show_traceback: Optional[bool] = None) -> Abort:
```
Comment (fubuloubu) [inline src/ape/exceptions.py:438]: make a classmethod for `Abort` and call it `from_ape_exception`?

**Review by fubuloubu — DISMISSED** [review]

**Pull request merged**